profiling/report.py

```diff
@@ -13,80 +12,162 @@
-df = (
-    pd.read_csv(StringIO(report), sep=" +", engine="python")
-    .rename(columns={"percall": "tottime_percall", "percall.1": "cumtime_percall"})
-    .sort_values(by="cumtime_percall", ascending=False)
-    .query("cumtime > 0.01")
-)
-df[["filename", "lineno", "function"]] = df.pop(
-    "filename:lineno(function)"
-).str.extract(r"(.*?):(.*?)\((.*)\)", expand=True)
-dt = datetime.datetime.fromtimestamp(prof_path.stat().st_mtime)
-df.index = pd.Series([dt.strftime("%Y-%m-%d %H:%M:%S")] * len(df), name="timestamp")
-print(df)
-df["label"] = df["filename"].str.extract(r"(\w+).py").squeeze() + "." + df["function"]
-df["event"] = [args.event] * len(df)
-report_path = root / "profiling/prof-report.csv"
-df.to_csv(report_path, mode="a", header=not report_path.exists())
-
-# Filter and plot the results
-df.plot.barh(y=["tottime_percall", "cumtime_percall"], x="label", figsize=(20, 10))
-plt.ylabel("")
-plt.gca().invert_yaxis()
-plt.tight_layout()
-plt.legend(loc="lower right")
-plt.savefig(root / "profiling/profiling.png")
-
-# Plot benchmark results
+
+def read_prof_as_dataframe(prof_path: str | Path) -> pd.DataFrame:
+    """Convert the ".prof" profiling result to a DataFrame."""
+
+    # Read the profiling results
+    sio = StringIO()
+    p = pstats.Stats(str(prof_path), stream=sio)
+    p.print_stats("pyrealm")
+    sio = StringIO(sio.getvalue())
+
+    # Convert to a DataFrame
+    df = (
+        pd.read_csv(sio, sep=" +", engine="python", skiprows=7)
+        .rename(columns={"percall": "tottime_percall", "percall.1": "cumtime_percall"})
+        .sort_values(by="cumtime_percall", ascending=False)
+        .query("cumtime > 0.01")  # drop insignificant functions
+        .dropna()
+    )
+
+    return df
+
+
+def process_df_columns(df: pd.DataFrame, path_filter: str = "") -> pd.DataFrame:
+    """Process the DataFrame columns including file name, function name, etc."""
+
+    # Split the filename, lineno, and function
+    df[["filename", "lineno", "function"]] = df.pop(
+        "filename:lineno(function)"
+    ).str.extract(r"(.*):(.*?)\((.*)\)", expand=True)
+
+    # Filter the paths by a regex pattern (usually the module name)
+    if path_filter:
+        df = df[df.filename.str.contains(path_filter)]
+
+    # Remove the common prefix of the filenames
+    i_diff = next(i for i, cs in enumerate(zip(*df.filename)) if len(set(cs)) > 1)
+    df["filename"] = [s[i_diff:] for s in df.filename]
+
+    # Add current time as timestamp
+    dt = datetime.datetime.fromtimestamp(prof_path.stat().st_mtime)
+    df["timestamp"] = dt.strftime("%Y-%m-%d %H:%M:%S")
+
+    # Add a unique label for each function for plotting
+    df["label"] = (
+        df["filename"].str.extract(r"(\w+).py").squeeze() + "." + df["function"]
+    )
+
+    # Add GitHub event name for information
+    df["event"] = args.event
+
+    return df
+
+
+def plot_profiling(df: pd.DataFrame, save_path: str | bool = False) -> None:
```

Comment:
We've not moved to Python 3.10 yet, so this needs to be `Union[str, bool]` - I'm also not sure why it can't just be `save_path: str = ""`. I'm not 100% keen on the Python `falsy` values for tests but it seems simpler here.

Reply:
I've passed `args` containing configs like `save_path` instead.